p4-16/psa/PSA.mdk

```diff
@@ -544,146 +546,168 @@ no payload.
-
-A PSA implementation provides a metadata bit `is_recirc` as the input
-metadata to the parser to indicate if a packet is a recirculation
-packet. The `is_recirc` bit is true if the packet is a recirculation
+### Packet Resubmission  { #sec-resubmit }
+
+Packet resubmission is a mechanism to repeat ingress processing on a
 packet.
 
-One possible implementation of the `is_recirc` bit is to set the bit
-based on the `ingress_port` metadata of the received packet. If the
-`ingress_port` metadata is equal to the dedicated recirculation port
-number, then the `is_recirc` bit is set.
+If a packet is resubmitted, it re-enters the ingress parser after the
+ingress pipeline processing. The `ingress_port` of the resubmitted
+packet is the same as the original packet, however, the
+`instance_type` of the resubmitted packet is changed to `RESUBMIT`.
+The resubmitted packet has the same header and payload as the original
+packet.
+
+The ingress parser is able to distinguish between the original packet
+and the resubmitted packet using the `instance_type` field in
+`ingress_parser_intrinsic_metadata_t`. The ingress parser is allowed
+to choose a different algorithm to parse the resubmitted packet.
+Similarly, the ingress pipeline can choose to process the resubmitted
+packet with different actions as opposed to the ones used to process
+the original packet.
+
+TBD: does PSA specify that a packet can be resubmitted at most once?
+
+PSA specifies that the resubmit operation can only be used in the
+ingress pipeline. The egress pipeline cannot resubmit packets.
+
+Packet resubmission is intended to increase the capacity and
+ flexibility of the packet processing pipeline. For example, because
+ the same packet is processed by the ingress pipeline multiple times,
+ it effectively increase the amount of operations on the packet by N
+ folds. Another use case is deploy multiple packet processing
+ algorithms on the same packet. For example, the original packet can
+ be parsed, modified and resubmitted in the first pass. Then, the
+ resubmitted packet can be parsed and modified using a different
+ algorithm.
+
+To facilitate communication between the first pass and the second pass
+through the ingress pipeline, the resubmission mechanism supports
+attaching optional metadata with the resubmitted packet. The metadata
+is generated during the first pass through the ingress pipeline, and
+used in the second pass.
+
+A PSA implementation provides the `resubmit` extern to specify the
+attached metadata. The `resubmit` extern provides a `add_metadata`
+method which accepts metadata with generic type `T`. In a PSA
+implementation, the metadata is prepended to the resubmitted packet.
+It is the responsibility of the programmer to parse the resubmitted
+packet correctly to extract the attached metadata. The attached
+metadata can be of type `header`, `header stack`, `header union` or
+`struct` of the above types. Invoking the `add_metadata` method
```

Comment:
how does this work with the serialization ideas? do we care to restrict to serializable data only? The scenario I'm thinking about is that the final packet after processing the resubmitted data might go out on the wire with that data as part of the header.

Reply:
I am going to rewrite this paragraph after today's discussion on #453. If the proposal in #453 is accepted, then we are not restricted to resubmit serializable data (header, header stack, header union, struct), but also enum, error, etc. The API to clone will be `pack` instead of `add_metadata`. If the user wants to send out enum, he/she has to pick an encoding for the enum/error, and handle the encoding at the receiving end.